source/core/EasyLogging.hpp

```diff
@@ -0,0 +1,307 @@
+            return *this;
+        }
+
+        /**
+         * @brief Manipulator for endl so that we can reset the values when a team is done logging
+         *
+         * @param manipulator
+         * @return Logger&
+         */
+        Logger &operator<<(std::ostream &(*manipulator)(std::ostream &)) {
+
+            typedef std::ostream &(*EndlManipulator)(std::ostream &);
+
+            // Compare the function pointers
+            if (manipulator == static_cast<EndlManipulator>(std::endl)) {
+                // Handle std::endl here
+                currentTeam = Team::NA;
+                currentLogLevel = LogLevel::DEBUG;
+                currentColor = Color::RESET;
+                std::cout << std::endl;
+            }
+
+            if (manipulator == endl) {
+
+                currentTeam = Team::NA;
+                currentLogLevel = LogLevel::DEBUG;
+                currentColor = Color::RESET;
+
+                std::cout
+                        << std::endl; //TODO: Might have to enable this so that we can have same line logging when endl is not used
+                metaPrinted = false;
+            }
+            return *this;
+        }
+
+        /**
+         * @brief Aye I used a template.
+         * @brief Function to log the value
+         *
+         * @tparam T
+         * @param value
+         * @return Logger&
+         *
+         * @TODO: Might have to change this so that we only break a team log when a new team is set. aka ensure that  logger << Team::TEAM_1 << "Hello" << "World" << endl; works in one line with one team print
+         */
+        template<typename T>
+        Logger &operator<<(const T &value) {
+
+
+            //TODO: Define when to log by loglevel comparison. Goal is to send it in as a flag in the CMakeLists.txt
+            if (currentLogLevel >= LOGLEVEL) {
+
+                // added additional flag in case one wants to compile without colors (or) if the terminal does not support colors
+#ifndef D_ANSI_COLOR_CODES
+                std::string colorStart = "\033[" + std::to_string(static_cast<int>(currentColor)) + "m";
+                std::string colorEnd = "\033[0m";
+#else
+                std::string colorStart = "";
+                std::string colorEnd = "";
+#endif
+                std::ostringstream logMessage;
+                logMessage << colorStart;
+                if (!metaPrinted) {
+                    logMessage << teamToString(currentTeam) << logToString(currentLogLevel);
+                    metaPrinted = true;
+                }
+
+                logMessage << value << colorEnd;
+                std::cout
+                        << logMessage.str(); // << std::endl;  //TODO: Might have to make enable this so that we can have same line logging when endl is not used
+            }
+
+            return *this;
+        }
+
+
+        static Logger log; /// Global log instance //TODO: Check if poluting the global namespace is a good idea??
```

Comment:
This line will create a static logger object for THIS compilation unit, but it will be different loggers across compilation units.  If you want to make sure there is only a single logger, you could do that by declaring it with:

```
static Logger & Log() {
  static Logger log;  // Creates this instance of log only when called the first time, but always uses same one.
  return log;  // Returns a consistent instance of log.
}
```

Of course then you will need to use `Logger::Log()` at the beginning of each log (instead of `Logger::log`)

Reply:
At this point I would encourage you to ALSO add another version of log that takes the arguments to print and doesn't require an endl at the end.  Specifically something like:

```
template <typename T, typename... EXTRA_Ts>
static Logger & Log(T && arg1, EXTRA_Ts &&... extra_args) {
  Log() << std::forward<T>(arg1);  // Log the first argument.
  if constexpr (sizeof...(EXTRA_Ts) == 0) {  // No arguments left.
    return Log() << Logger::endl;  // Trigger a flush.
  } else {
    return Log(std::forward<EXTRA_Ts>(extra_args)...);  // Log remaining arguments.
  }
}
```

Reply:
I have made both changes. 

I have also added a macro 

`#define log Log()`

where we can still choose to call it by 
`Logger::log`

I am not sure if this is a poor choice on my end. But I like to syntax of using `Logger::log` without the parenthesis.
I understand this is purely syntactical, but I am a little opposed to changing it unless it is a bad practice.